overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
+
+	// put the snaps in the group
+	grp.Snaps = action.AddSnaps
+	// update the modified groups in state
+	newAllGrps, err := patchQuotas(st, updatedGrps...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return grp, newAllGrps, nil
+}
+
+type quotaControlArgs struct {
+	St          *state.State
+	AllGrps     map[string]*quota.Group
+	Action      *QuotaControlAction
+	Meter       progress.Meter
+	PerfTimings *timings.Timings
+}
+
+func doRemoveQuota(args quotaControlArgs) error {
+	st := args.St
+	action := args.Action
+	allGrps := args.AllGrps
+
+	// make sure the group exists
+	grp, ok := allGrps[action.QuotaName]
+	if !ok {
+		return fmt.Errorf("group %q does not exist", action.QuotaName)
+	}
+
+	// make sure some of the options are not set, it's an internal error if
+	// anything other than the name and action are set for a removal
+	if action.ParentName != "" {
+		return fmt.Errorf("internal error, ParentName option does not make sense with remove action")
+	}
+
+	if len(action.AddSnaps) != 0 {
+		return fmt.Errorf("internal error, AddSnaps option does not make sense with remove action")
+	}
+
+	if action.MemoryLimit != 0 {
+		return fmt.Errorf("internal error, MemoryLimit option does not make sense with remove action")
+	}
+
+	// XXX: remove this limitation eventually
+	if len(grp.SubGroups) != 0 {
+		return fmt.Errorf("cannot remove quota group with sub-groups, remove the sub-groups first")
+	}
+
+	// if this group has a parent, we need to remove the linkage to this
+	// sub-group from the parent first
+	if grp.ParentGroup != "" {
+		// the parent here must exist otherwise AllQuotas would have failed
+		// because state would have been inconsistent
+		parent := allGrps[grp.ParentGroup]
+
+		// ensure that the parent group of this group no longer mentions this
+		// group as a sub-group - we know that it must since AllQuotas validated
+		// the state for us
+		if len(parent.SubGroups) == 1 {
+			// this group was an only child, so clear the whole list
+			parent.SubGroups = nil
+		} else {
+			// we have to delete the child but keep the other children
+			newSubgroups := make([]string, 0, len(parent.SubGroups)-1)
+			for _, sub := range parent.SubGroups {
+				if sub != action.QuotaName {
+					newSubgroups = append(newSubgroups, sub)
+				}
+			}
+
+			parent.SubGroups = newSubgroups
+		}
+
+		allGrps[grp.ParentGroup] = parent
+	}
+
+	// now delete the group from state - do this first for convenience to ensure
+	// that we can just use SnapServiceOptions below and since it operates via
+	// state, it will immediately reflect the deletion
+	delete(allGrps, action.QuotaName)
+
+	// make sure that the group set is consistent before saving it - we may need
+	// to delete old links from this group's parent to the child
+	if err := quota.ResolveCrossReferences(allGrps); err != nil {
+		return fmt.Errorf("cannot remove quota %q: %v", action.QuotaName, err)
+	}
+
+	// now set it in state
+	st.Set("quotas", allGrps)
```

Comment:
I think this is ok, even if we reboot before all changes in the slice unit are done. The state is only updated on disk in Unlock(), unless ensureSnapServicesForGroup unlocks the state too. Otherwise we could end up in a situation when test is restarted and the group is already gone from the state.